Add module builder

pedorich-n opened the pull request:
This PR adds a module interface to the builder derivation. This brings benefits such as strict options definition, better type-checks, and flexibility for future improvements or modifications. 
I left the old builder in place for backward compatibility. In the future, it could be removed, and an "adapter" could be installed to convert old inputs into the new module system and maintain backward compatibility.

The module is inspired by @tomeon's [nix-openwrt-imagebuilder-module](https://git.sr.ht/~tomeon/nix-openwrt-imagebuilder-module/tree/6c2a5fc19c5a6fd532b99c69e5b4ce14adfd80fe/item/module.nix) project, but it's structured differently and the module itself is simpler. Please let me know if you feel it isn't fair to your work. 

Opening this as a draft to gather opinions. 
If everything is fine I will add module documentation and update README to tell users to use `build-module` in a follow-up PR to not pollute this one. 

I also moved examples into a dedicated folder, since we have 4 of them now.

Closes #22

pedorich-n commented on builder/module-options.nix:
```diff
@@ -0,0 +1,272 @@
+      };
+    };
+
+    sumsFileSha256 = lib.mkOption {
```
It's hard to come up with a good name for this one 😅

SuperSandro2000 replied:
We probably want to use Hash instead of Sha256 for future proofness.

pedorich-n commented on builder/module-options.nix:
```diff
@@ -0,0 +1,272 @@
+  # TODO: get rid of `pkgs`
+  profiles = import ../profiles.nix { pkgs = null; inherit (config) release; };
+  detectedProfile = profiles.identifyProfile config.hardware.profile;
+  # TODO: move arch to profiles
+  detectedArch = hashedTarget.packagesArch;
```
`arch` is already a property in `profiles.json`. I feel like it's more fitting to get this info from there, than from the feeds, since it's a property of a specific board/target

pedorich-n reviewed (commented):
Also, I would love for people to test this, maybe I missed some corner cases.
In my tests, I tried building with packages, folders, and individual files. I haven't tried disabling services, removing packages or anything else.

astro commented:
Cc: @ehmry @adamcstephens @makisekurisu

ehmry commented:
I'd like to review but I haven't been using the imagebuilder.

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+    inherit pkgs;
+    inherit (config) release feedsSha256;
+    inherit (config.hardware) target;
+    variant = config.hardware.subtarget;
```
Shouldn't we name variant subtarget here, too or at least add a TODO comment to rename it?

pedorich-n replied:
I didn't want this PR to turn into a whole repository rewrite. But I'll add a TODO.

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+    inherit (config.hardware) target;
+    variant = config.hardware.subtarget;
+    packagesArch = config.hardware.arch;
+    sha256 = config.sumsFileSha256;
```
```suggestion
    hash = config.sumsFileHash;
```

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+    let
+      package = allPackages.${pname};
+    in
+    lib.optional (package.type == "real") ''[ -e "packages/${package.filename}" ] || ln -s ${package.file} "packages/${package.filename}"'';
```
```suggestion
    lib.optional (package.type == "real") ''
      [ -e "packages/${package.filename}" ] || ln -s ${package.file} "packages/${package.filename}"
    '';
```

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+    )
+  );
+
+  mkCopyStatement = file: ''(
+     target_rel=${lib.escapeShellArg file.target}
+     target="files/''${target_rel#/}"
+     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
+        mkdir -p $(dirname "''${target}")
+     ''}
+     cp --recursive --no-preserve=all "${file.source}" "''${target}"
+    )'';
```
```suggestion
  mkCopyStatement = file: ''
    (
     target_rel=${lib.escapeShellArg file.target}
     target="files/''${target_rel#/}"
     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
        mkdir -p "$(dirname "''${target}")"
     ''}
     cp -r --no-preserve=all "${file.source}" "''${target}"
    )
  '';
```

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+  src =
+    let
+      inherit (pkgs.stdenv.hostPlatform) uname;
+      imageBuilderPrefix = "openwrt-imagebuilder-${if config.release == "snapshot" then "" else "${config.release}-"}";
```
```suggestion
      imageBuilderPrefix = "openwrt-imagebuilder-${lib.optionalString (config.release != "snapshot") "${config.release}-"}";
```

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+    in
+    if matches != [ ]
+    then builtins.getAttr (builtins.elemAt matches 0) variantFiles
+    else throw "No valid image builder found!";
```
Can we easily name which one we searched for to make debugging easier?

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,169 @@
+in
+{
+
+  ###### implementation
```
```suggestion
```
lets no copy paste legacy comments from nixos

SuperSandro2000 commented on builder/module-docs.nix:
```diff
@@ -0,0 +1,34 @@
+    config._module = {
+      check = false; # Don't check for unset variables
+      args = {
+        inherit pkgs lib; # Provide pkgs and lib to modules
```
From my experience with nüschtos search you generally don't need to provide lib.

SuperSandro2000 commented on builder/module-docs.nix:
```diff
@@ -0,0 +1,34 @@
+
+  makeOptionsDoc = module: (pkgs.nixosOptionsDoc {
+    options = optionsFor module;
+    transformOptions = options: builtins.removeAttrs options [ "declarations" ];
```
We probably want to filter declarations separately and also provide the optionsJSON to easily integrate it into nüschtos search.

pedorich-n replied:
I'm not familiar with "nüschtos search". Do you have an example of how the module should look like?

SuperSandro2000 replied:
It should look similar to this https://github.com/NuschtOS/search/blob/3051be7f403bff1d1d380e4612f0c70675b44fc9/nix/wrapper.nix#L4-L17

SuperSandro2000 commented on builder/module-options.nix:
```diff
@@ -0,0 +1,272 @@
+  latestRelease = import ../latest-release.nix;
+
+  releaseHashes = import ../hashes/${config.release}.nix;
+  hashedTarget = releaseHashes.targets.${config.hardware.target}.${config.hardware.subtarget};
+  hashedFeeds = releaseHashes.packages.${config.hardware.arch};
```
We should probably add a few or's here to make errors easier to debug.

SuperSandro2000 commented on builder/module-options.nix:
```diff
@@ -0,0 +1,272 @@
+  };
+in
+{
+  ###### interface
```
```suggestion
```

astro commented:
I like this. Please feel encouraged to follow through.

pedorich-n commented:
Thanks for your encouragement, @astro !

I've updated the docs module, exposing both markdown and json. At a certain iteration, I realized it looked almost like https://github.com/astro/microvm.nix/blob/2ae08de8e8068b00193b9cfbc0acc9dfdda03181/pkgs/doc.nix, so I just copied most of the stuff from there 😅 
@SuperSandro2000 please have a look.

I also added support for "separate Kmods" and tested the build with both "old" (23.05.5) and "new" (24.10.0-rc4) versions. Builds fine.

pedorich-n commented:
Happy New Year everybody!

I've added a new package to generate/update module docs and a GitHub flow to do it from GHA. 
I've also updated the documentation with a minimal example and a recommendation to use the builder module over the builder function.

I think this PR is ready for a proper review now.

pedorich-n commented:
Is there any chance I could get a review @SuperSandro2000 @astro ? 👀

SuperSandro2000 commented on builder/module-config.nix:
```diff
@@ -0,0 +1,178 @@
+      postPatch = ''
+        patchShebangs scripts staging_dir/host/bin
+        substituteInPlace rules.mk \
+          --replace "SHELL:=/usr/bin/env bash" "SHELL:=${pkgs.runtimeShell}" \
+          --replace "/usr/bin/env true" "${pkgs.coreutils}/bin/true" \
+          --replace "/usr/bin/env false" "${pkgs.coreutils}/bin/false"
```
Should we use --replace-fail here?

SuperSandro2000 commented on builder/module-options.nix:
```diff
@@ -0,0 +1,301 @@
+    release = lib.mkOption {
+      type = lib.types.nonEmptyStr;
+      default = latestRelease;
+      defaultText = lib.literalMD "Value of `latest-release.nix`";
```
```suggestion
      defaultText = "Value of `latest-release.nix`";
```
literalMD doesn't exist anymore in unstable

SuperSandro2000 commented on docs/module-options.md:
I would not commit this file and generate it on the fly for github pages or use https://github.com/nuschtos/search

SuperSandro2000 commented:
I would really need to try this to know fore sure if it works perfectly but I am not keen on bricking my router right now in case of any issues 😓

astro commented on builder/module-config.nix:
```diff
@@ -0,0 +1,178 @@
+    )
+  );
+
+  mkCopyStatement = file: ''
+    (
+     target_rel=${lib.escapeShellArg file.target}
+     target="files/''${target_rel#/}"
+     ${lib.optionalString (lib.pathIsRegularFile file.source) ''
+        mkdir -p "$(dirname "''${target}")"
+     ''}
+     cp -r --no-preserve=all "${file.source}" "''${target}"
+    )
+  '';
+
+  copyFiles = pkgs.writeScript "copy-openwrt-files" (
+    lib.concatLines (builtins.map mkCopyStatement config.files)
+  );
```
I actually prefer having once-used definitions inline for readability. The indentation isn't very deep here so I think it would be fine:

```suggestion
  copyFiles = pkgs.writeScript "copy-openwrt-files" (
    lib.concatMapStrings (file: ''
      (
        target_rel=${lib.escapeShellArg file.target}
        target="files/''${target_rel#/}"
        ${lib.optionalString (lib.pathIsRegularFile file.source) ''
          mkdir -p "$(dirname "''${target}")"
        ''}
        cp -r --no-preserve=all "${file.source}" "''${target}"
      )
    '') config.files
  );
```

astro reviewed (changes requested):
As you already said, if I merge the modules like this we're creating duplicate functionality in this project. Any intermediate PR would have to do twice the work.

That's why I am going to wait for the removal of the old code and the addition of a few shims.

SuperSandro2000 commented:
Me and @m4rc3l have some interest in using this for future infrastructure. Hopefully we find some time bringing this forward together.

pedorich-n commented:
I'm sorry I haven't been active. I really want to see this working, but I struggled to find the willpower to get back to it.
The changes in this PR have grown quite large, and being asked to completely replace existing code within the same PR became overwhelming for me.

I'd still love to help if we can find a way to contribute incrementally.

MarcelCoding commented:
I've successfully deployed my testing AP using this PR in https://gitea.c3d2.de/zentralwerk/network/src/branch/openwrt-module. (I've used my fork to get up-to-date hashes)
